plugin/pkg/scheduler/core/generic_scheduler.go

```diff
@@ -423,11 +484,350 @@ func EqualPriorityMap(_ *v1.Pod, _ interface{}, nodeInfo *schedulercache.NodeInf
+			podIndexesForAffinity = append(podIndexesForAffinity, i)
+			if fits, _, _ = interPodAffinityPredicate(pod, meta, nodeInfoCopy); fits {
+				affinityCanBeSatisfied = true
+				break
+			}
+		}
+		if !affinityCanBeSatisfied {
+			glog.Errorf("Bobby: affinity cannot be satisfied.")
+			// Even after adding all the victims, affinity rules cannot be satisfied.
+			// The failure may be related to anti-affinity to higher priority pods or
+			// affinity to pods that do not exist in this topology.
+			return nil, false
+		}
+		// So far we have found all the pods that satisfy the affinity rules, but
+		// this is not the minimum set. We start removing the extra ones in reverse
+		// order we added them, which means remove lowest priority first. We don't
+		// bother with the very last one we added. It's necessary to satisfy the
+		// affinity predicate.
+		for i := len(podIndexesForAffinity) - 2; i >= 0; i-- {
+			victim := potentialVictims.Items[podIndexesForAffinity[i]].(*v1.Pod)
+			removePod(victim)
+			glog.Errorf("Bobby: remove pod %v.", victim.Name)
+			if fits, _, _ = interPodAffinityPredicate(pod, meta, nodeInfoCopy); !fits {
+				// This pod is needed. Add it back and remove it from the list of victims.
+				addPod(victim)
+				potentialVictims.Items = append(potentialVictims.Items[:podIndexesForAffinity[i]], potentialVictims.Items[podIndexesForAffinity[i]+1:]...)
+			}
+		}
+		// We have found the minimum set of pods needed to satisfy affinity rules.
+		// Let's see if other predicates are happy.
+		if fits, _, _ = podFitsOnNode(pod, meta, nodeInfoCopy, fitPredicates, nil); !fits {
+			return nil, false
+		}
+	}
+	victims := []*v1.Pod{}
+	// Try to reprieve as many pods as possible starting from the highest priority one.
+	for _, p := range potentialVictims.Items {
+		lpp := p.(*v1.Pod)
+		addPod(lpp)
+		if fits, _, _ := podFitsOnNode(pod, meta, nodeInfoCopy, fitPredicates, nil); !fits {
+			removePod(lpp)
+			victims = append(victims, lpp)
+			glog.V(5).Infof("Pod %v is a potential preemption victim on node %v.", lpp.Name, nodeInfo.Node().Name)
+		}
+	}
+	return victims, true
+}
+
+// nodesWherePreemptionMightHelp returns a list of nodes with failed predicates
+// that may be satisfied by removing pods from the node.
+func nodesWherePreemptionMightHelp(pod *v1.Pod, nodes []*v1.Node, failedPredicatesMap FailedPredicateMap) []*v1.Node {
+	potentialNodes := []*v1.Node{}
+	for _, node := range nodes {
+		unresolvableReasonExist := false
+		failedPredicates, found := failedPredicatesMap[node.Name]
+		// If we assume that scheduler looks at all nodes and populates the failedPredicateMap
+		// (which is the case today), the !found case should never happen, but we'd prefer
+		// to rely less on such assumptions in the code when checking does not impose
+		// significant overhead.
+		if found {
+			for _, failedPredicate := range failedPredicates {
+				switch failedPredicate {
+				case
+					predicates.ErrNodeSelectorNotMatch,
```

Comment:
does this mean that if new errors not listed here occur, we might think preempting will help and delete pods incorrectly? would it not be better to enumerate the errors for which preemption would help?

Reply:
No, it means that we may try to preempt pods without any success. Not adding new error types that cannot be resolved by preemption will waste CPU cycles, but won't affect correctness.